### PR TITLE
Reduce allocations 2

### PR DESCRIPTION
**Pull request by iamcarbon**
- Eliminates a (byte) cast operation
- Eliminates DateToken allocations
- Eliminates ByteArrayBuilder allocations
- Eliminates pinning in Unix2DosFilter

### CODE REVIEW

**Comment by iamcarbon** [comment]
@jstedfast Ready for review.

File: MimeKit/Utils/DateUtils.cs [hunk 1]
```diff
@@ -353,7 +355,7 @@ static IEnumerable<DateToken> TokenizeDate (byte[] text, int startIndex, int len
 			yield break;
 		}
 
-		static bool TryParseStandardDateFormat (IList<DateToken> tokens, byte[] text, out DateTimeOffset date)
+		static bool TryParseStandardDateFormat (List<DateToken> tokens, byte[] text, out DateTimeOffset date)
```
Comment (iamcarbon) [inline MimeKit/Utils/DateUtils.cs:358]: Eliminates possible virtual dispatch.

File: MimeKit/Utils/ByteArrayBuilder.cs [hunk 2]
```diff
@@ -28,26 +28,17 @@
 using System.Buffers;
 
 namespace MimeKit.Utils {
-	internal sealed class ByteArrayBuilder : IDisposable
+	internal ref struct ByteArrayBuilder
```
[the remark this answers is not in the document]
Comment (jstedfast) [reply MimeKit/Utils/ByteArrayBuilder.cs:31]: With the using statement still work if this no longer implements IDisposable? I didn't think it would...

**Comment by jstedfast** [comment]
I wasn't sure about the Unix2DosFilter changes, so I wrote up a benchmark...

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=6.0.200-preview.22055.15
  [Host]     : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT


|                     Method |     Mean |    Error |   StdDev |   Median |
|--------------------------- |---------:|---------:|---------:|---------:|
|     Unix2Dos_LoremIpsumDos | 99.63 ns | 1.972 ns | 5.498 ns | 96.94 ns |
|    Unix2Dos_LoremIpsumUnix | 97.86 ns | 1.701 ns | 1.591 ns | 97.85 ns |
|  NewUnix2Dos_LoremIpsumDos | 97.07 ns | 1.394 ns | 1.236 ns | 97.37 ns |
| NewUnix2Dos_LoremIpsumUnix | 96.82 ns | 0.617 ns | 0.515 ns | 96.80 ns |

**Pull request merged**